### DISCUSSION TITLE
Headless / Sessions mode

### DISCUSSION THREAD

**Discussion by drunlade** [webui-dev/webui#429] [Ideas]
I was interested in using this as a means of implementing a web application since there isn't much around that does what you're doing without being some sort of node.js monstrosity. However it's "single user" focus makes this difficult (if not impossible).

I'd be very interested in a headless mode (where you don't have to invoke a local WebView to have the application run) and instead just have someone connect via whatever browser they have.

Coupled with this, it'd be nice to have the means of running multiple sessions, something along the lines of:

```go
webui.HandleClient(
  func(c *webui.Client){
    s := webui.NewSession(c)
    s.NewWindow()
    s.ShowHeadless("index.html")
  }
}
```

I'm aware this isn't the original aim of this project, but it'd turn it into the go-to method of creating proper web applications without some massive node stack behind the scenes.

**Comment by hassandraga** [comment]
If I understand correctly your explanation, I guess you are looking for an event of a new connection from an external web browser, please correct me if I'm wrong.

I guess you can do it like this:

1. Set `webui_set_timeout(0)` to make window UI always available
2. Use `webui_get_url(win)` to get URL
3. Set `bind("", your_event_handler)` to receive new client events
4. When a user connects from any browser using the URL you will receive `e->event_type == WEBUI_EVENT_CONNECTED`

More details in the example: https://github.com/webui-dev/webui/blob/main/examples/C/public_network_access/main.c

About the headless, I still not fully understand the goal.

**Reply by drunlade** [reply]
Would the above suggestion work if two different clients connected at the same time? I got the impression that internally webui has one state, maybe I'm mistaken.

> About the headless, I still not fully understand the goal.

It seemed that if I don't create a window that is rendered on the local machine, webui immediately exits. Headless mode would prevent this by skipping local rendering/display. However, I think your suggestion of `webui_set_timeout(0)` is the way to avoid that? Does that stop webui from exiting when there are no rendered windows currently? I'm wondering if perhaps the Go wrapper isn't exposing everything I need here.

My goal is:

* webui is running, has no open windows yet
* Client connects
* Create a new window that is bound to that Client alone
* Second Client connects
* Create a second window, bound to the second Client alone
* Both Clients close their windows/disconnect
* There are now no windows, but webui remains running, waiting for a new connection to spawn a new window

**Reply by hassandraga** [reply]
> work if two different clients connected at the same time?

WebUI is not multi clients, it was before 2.3.x but we removed this feature to focus on GUI functionality and not framework. Now, every UI have one unique client. You will still receive new client's connection events, but the new clients will receive an error `Access Denied`.

> I got the impression that internally webui has one state

Yes.

> It seemed that if I don't create a window that is rendered on the local machine, webui immediately exits

Yes, except if you set `webui_set_timeout(0)` which make WebUI waits forever.

> Create a new window that is bound to that Client alone

No, if you create a new window, you may need to manually redirect the client to the new window's URL.

In general, WebUI is a GUI library, every window work with a single client only.
However... multi-clients feature is a good thing to have in the future... But in the current version it's not available for now.

**Reply by hassandraga** [reply]
Poll: https://github.com/webui-dev/webui/discussions/434

**Reply by drunlade** [reply]
Thanks, that clears it all up, and I appreciate the poll.

I could potentially work around the issue by having a router that accepts clients and spawns separate application instances and routes the client connection to them. However this would probably be easier if webui could be told to use a unix socket instead of a port.
